libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_pthread.c

```diff
@@ -364,6 +411,182 @@ TEST( Full_POSIX_PTHREAD, pthread_create_join )
+            TEST_FAIL_MESSAGE( "Failed to release the barrrier for detached thread." );
+        }
+    }
+
+    /* Destroy the barrier. */
+    ( void ) pthread_barrier_destroy( &xBarrier );
+}
+
+/*-----------------------------------------------------------*/
+
+TEST( Full_POSIX_PTHREAD, pthread_detach_self )
+{
+    int iStatus = 0;
+    pthread_t xNewThread = NULL;
+    pthread_barrier_t xBarrier = { 0 };
+
+    /*Initialize a pthread barrier for two threads. */
+    iStatus = pthread_barrier_init( &xBarrier, NULL, 2U );
+    TEST_ASSERT_EQUAL_INT( 0, iStatus );
+
+    if( TEST_PROTECT() )
+    {
+        /* Create a thread with default attributes. By default thread is created as joinable.
+         * Thread will detach by itself after running.
+         */
+        iStatus = pthread_create( &xNewThread, NULL, prvSelfDetachThread, ( void * ) &xBarrier );
+        TEST_ASSERT_EQUAL_INT_MESSAGE( 0,
+                                       iStatus,
+                                       "Thread creation failed!" );
+
+        /* Wait on the barrier for thread to detach by itself.*/
+        iStatus = pthread_barrier_wait( &xBarrier );
+
+        if( ( iStatus != 0 ) && ( iStatus != PTHREAD_BARRIER_SERIAL_THREAD ) )
+        {
+            TEST_FAIL_MESSAGE( "Failed to release the barrrier for detached thread." );
+        }
+
+        /* Attempt to join a detached thread. This should fail with EDEADLK. */
+        iStatus = pthread_join( xNewThread, NULL );
+        TEST_ASSERT_EQUAL_INT( EDEADLK, iStatus );
+
+        /* Trying to detach the thread again should return EINVAL. */
+        iStatus = pthread_detach( xNewThread );
+        TEST_ASSERT_EQUAL_INT( EINVAL, iStatus );
+
+        /* Wait in the barrier again, which will release prvSelfDetachThread() and the detached thread should exit. */
+        iStatus = pthread_barrier_wait( &xBarrier );
+
+        if( ( iStatus != 0 ) && ( iStatus != PTHREAD_BARRIER_SERIAL_THREAD ) )
+        {
+            TEST_FAIL_MESSAGE( "Failed to release the barrrier for detached thread." );
+        }
+    }
+
+    /* Destroy the barrier. */
+    ( void ) pthread_barrier_destroy( &xBarrier );
+}
+
+/*-----------------------------------------------------------*/
+
+TEST( Full_POSIX_PTHREAD, pthread_detach_after_running )
+{
+    int iStatus = 0;
+    pthread_t xNewThread = NULL;
+    volatile int iNumberToSquare = 5;
+    struct timespec xDelay =
+    {
+        .tv_sec  = 0,
+        .tv_nsec = posixtestPTHREAD_DETACHED_WAIT_NANOSECONDS
+    };
+
+    if( TEST_PROTECT() )
+    {
+        iStatus = pthread_create( &xNewThread, NULL, prvComputeSquareThread, ( void * ) &iNumberToSquare );
+        TEST_ASSERT_EQUAL_INT_MESSAGE( 0,
+                                       iStatus,
+                                       "Thread creation failed!" );
+
+        /* Wait some time for prvComputeSquareThread to exit and ensure it completed
+         * the calculation. */
+        nanosleep( &xDelay, NULL );
```

Comment:
Using sleep for synchronization is not a good idea. Can we use barrier as in other tests?

Reply:
Removed sleep synchronization. However using pthread_barrier alone will not make sure the thread has suspended. Changed the test thread to use a higher priority than running thread to ensure the thread runs to suspension after releasing from barrier.

Reply:
We can probably avoid this dynamic priority altering by creating the new task with priority `currentTaskPriority + 1`?

Reply:
As discussed task which runs the test can run at max priority level in which case two threads will have same priority. To avoid this case, priority is altered temporarily.